File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/input/metrics/AddQuotDel.java
package de.maibornwolff.codecharta.importer.scmlogparser.input.metrics;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;

public class AddQuotDel implements Metric {
    private long accumulatedNumberOfLinesAdded = 0;
    private long accumulatedNumberOfLinesDeleted = 0;

    @Override
    public String description() {
        return "Relative Code Churn: Quotient of addition churn and deletion churn.";
    }

    @Override
    public String metricName() {
        return "add_quot_deletion";
    }

    @Override
    public Number value() {
        if (accumulatedNumberOfLinesDeleted > 0) {
            return 100 * accumulatedNumberOfLinesAdded / accumulatedNumberOfLinesDeleted;
        }
        return 100 * accumulatedNumberOfLinesAdded;
    }

    @Override
    public void registerModification(Modification modification) {
        accumulatedNumberOfLinesAdded += modification.getAdditions();
        accumulatedNumberOfLinesDeleted += modification.getDeletions();
    }
}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/metrics/AddQuotDel.kt
package de.maibornwolff.codecharta.importer.scmlogparser.input.metrics

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification

class AddQuotDel : Metric {
    private var accumulatedNumberOfLinesAdded: Long = 0
    private var accumulatedNumberOfLinesDeleted: Long = 0

    override fun description(): String {
        return "Relative Code Churn: Quotient of addition churn and deletion churn."
    }

    override fun metricName(): String {
        return "add_quot_deletion"
    }

    override fun value(): Number {
        return when {
            accumulatedNumberOfLinesDeleted > 0 ->
                100 * accumulatedNumberOfLinesAdded / accumulatedNumberOfLinesDeleted
            else ->
                100 * accumulatedNumberOfLinesAdded
        }
    }

    override fun registerModification(modification: Modification) {
        accumulatedNumberOfLinesAdded += modification.additions
        accumulatedNumberOfLinesDeleted += modification.deletions
    }
}
